lib/functions/OverloadNode.ts

```diff
@@ -0,0 +1,90 @@
+import type * as E from '../expressions';
+import type { ArgumentType } from './Core';
+
+export type SearchStack = OverloadNode[];
+
+/**
+ * Maps argument types on their specific implementation in a tree like structure.
+ */
+export class OverloadNode {
+  private implementation?: E.SimpleApplication | undefined;
+  private readonly subTrees: Record<string, OverloadNode>;
+
+  public constructor() {
+    this.implementation = undefined;
+    this.subTrees = {};
+  }
+
+  /**
+   * Adds an overload to the tree structure considering this as the tree's root.
+   * @param argumentTypes a list of ArgumentTypes that would need to be provided in the same order to
+   * get the implementation.
+   * @param func the implementation for this overload.
+   */
+  public addOverload(argumentTypes: ArgumentType[], func: E.SimpleApplication): void {
+    this._addOverload([ ...argumentTypes ], func);
+  }
+
+  private _addOverload(argumentTypes: ArgumentType[], func: E.SimpleApplication): void {
+    const argument = argumentTypes.shift();
+    if (!argument) {
+      this.implementation = func;
+      return;
+    }
+    const str = argument;
+    if (!this.subTrees[str]) {
+      this.subTrees[str] = new OverloadNode();
+    }
+    this.subTrees[str]._addOverload(argumentTypes, func);
+  }
+
+  /**
+   * @param arg term to try and match to possible overloads of this node.
+   * @returns SearchStack a stack with top element the next node that should be asked for implementation or overload.
+   */
+  public getSubTreeWithArg(arg: E.TermExpression): SearchStack {
+    // The numbers here aren't magic.
+    // They just provide a priority, a way to tell what type is more concrete and should be looked at first.
```

Comment:
Could you enumerate these types? Because it's not clear at the moment what these types are, and why you should look at some of them first.

Reply:
What do you mean?
Also I wouldn't give to much attention to these numbers as they should disappear the moment we implement the auto cast feature.

Reply:
I'm referring to this:
> what type is more concrete

Reply:
You mean just in the comment? 
Something like `The numbers a priority, a way to tell what type is more concrete and should be looked at first. Order of concreteness: 1. the type (eg: 'string'), 2. termType (eg: literal) and 3. a term.`

Reply:
Yes, something like that indeed.
But perhaps you could include an `@see` reference to a place where the reader can find more information about what type, termType and term mean in the context of sparqlee (because I don't fully understand it atm).